### PR TITLE
Feature/issue 699 cout cerr

### PR DESCRIPTION
**Pull request by syclik**
#### Summary:

Removes references of `std::cout` and `std::cerr` from all of Stan source. In places where `std::cout` was used, I added extra arguments taking a pointer to an `std::ostream`.

`src/stan/command/print.hpp` and `src/stan/command/command.hpp` were moved to CmdStan and there's a corresponding pull request there for this issue.

#### Intended Effect:

Removes `std::cout` and `std::cerr` from our code base. 

#### How to Verify:

grep for `std::cout` and `std::cerr`. All tests pass.

#### Side Effects:

Lots of extra arguments, but this should not affect any behavior. This is going to be a little painful for rstan and pystan, but I've gotten cmdstan working, so I'll help out.

#### Documentation:

I didn't add additional API doc, but if you think I need it before it goes in, let me know.

#### Reviewer Suggestions: 

@bob-carpenter

### CODE REVIEW

**Comment by betanalpha** [comment]
Is there a reason for this branch in light of https://github.com/stan-dev/stan/tree/feature/issue-1361-writer_callback?
All of this effort (especially with the interfaces) will be duplicated, and the inevitable merge conflict will be a giant
mess.

**Comment by syclik** [comment]
@betanalpha, I realize this effort was partially duplicated, but this fix is a lot simpler than the one on that branch. That branch has a lot more going on and it isn't ready for a pull request, as far as I can tell. 

This one is ready to go; there are only a few changes to the code base. The rest of the changes are to get the tests working.

Regarding the other branch, I'll help get that in; I'd like to split that up into multiple pull requests so it's manageable.

**Comment by betanalpha** [comment]
I understand that we’re waiting on the writer callback but I don’t get what
short-term problems this fixes.  It’s just going to add work for the interface
devs + the refactoring devs.

**Comment by syclik** [comment]
Here is what it fixes:
- long-term problem: `std::cout` and `std::cerr` have unexpected behavior through the interfaces. In the worst case, it crashes.
- arbitrary problem: this is holding us up for v2.7.0 release

Well, this is a moot point until I can get the performance tests working again.

**Comment by betanalpha** [comment]
Yeah, but the same long-term problem gets fixed by the writer callback
which wouldn’t require a bunch of new code to be written in the meantime.
Was this tagged for 2.7.0 arbitrarily or does it have a real immediate effect?

File: src/stan/model/util.hpp [hunk 1]
```diff
@@ -452,28 +439,26 @@ namespace stan {
...
+      stan::math::hessian_times_vector(model_functional<M>(model, msgs),
+                                       x, v, f, hess_f_dot_v);
     }
 
     template <class M>
     void grad_tr_mat_times_hessian(const M& model,
-                                   const Eigen::Matrix<double, Eigen::Dynamic, 1>& x,
-                                   const Eigen::Matrix<double, Eigen::Dynamic, Eigen::Dynamic>& X,
-                                   Eigen::Matrix<double, Eigen::Dynamic, 1>& grad_tr_X_hess_f,
+                                   const Eigen::Matrix
```
Comment (bob-carpenter) [inline src/stan/model/util.hpp]: one line as in

```
void grad_tr_mat_times_hessian(const M& model,
                                   const Eigen::Matrix<double, Eigen::Dynamic, 1>& x,
                                   const Eigen::Matrix<double, Eigen::Dynamic, Eigen::Dynamic>& X,
                                   Eigen::Matrix<double, Eigen::Dynamic, 1>& grad_tr_X_hess_f,
```

File: src/stan/services/init/init_adapt.hpp [hunk 2]
```diff
@@ -31,27 +33,31 @@ namespace stan {
...
                       stan::services::categorical_argument* adapt,
-                      const Eigen::VectorXd& cont_params) {
-
-        double delta = dynamic_cast<stan::services::real_argument*>(adapt->arg("delta"))->value();
-        double gamma = dynamic_cast<stan::services::real_argument*>(adapt->arg("gamma"))->value();
-        double kappa = dynamic_cast<stan::services::real_argument*>(adapt->arg("kappa"))->value();
-        double t0    = dynamic_cast<stan::services::real_argument*>(adapt->arg("t0"))->value();
+                      const Eigen::VectorXd& cont_params,
+                      std::ostream* o) {
+        double delta = dynamic_cast<stan::services::real_argument*>
```
Comment (bob-carpenter) [inline src/stan/services/init/init_adapt.hpp]: Start = on its own line.

File: src/test/unit/io/mcmc_writer_test.cpp [hunk 3]
```diff
@@ -37,7 +39,8 @@ TEST(StanIoMcmcWriter, write_sample_names) {
   typedef boost::ecuyer1988 rng_t;
   rng_t base_rng(0);
   
-  stan::mcmc::adapt_diag_e_nuts<io_example_model_namespace::io_example_model, rng_t> sampler(model, base_rng, 0);
+  stan::mcmc::adapt_diag_e_nuts<io_example_model_namespace::io_example_model, rng_t>
+    sampler(model, base_rng, 0, 0);
   sampler.seed(real);
```
Comment (bob-carpenter) [inline src/test/unit/io/mcmc_writer_test.cpp]: Shouldn't the seed be an integer?

File: src/test/unit/io/stan_csv_reader_test.cpp [hunk 4]
```diff
@@ -68,7 +68,7 @@ TEST_F(StanIoStanCsvReader,read_metadata1) {
 }
 TEST_F(StanIoStanCsvReader,read_header1) {
   Eigen::Matrix<std::string, Eigen::Dynamic, 1> header;
-  EXPECT_TRUE(stan::io::stan_csv_reader::read_header(header1_stream, header));
+  EXPECT_TRUE(stan::io::stan_csv_reader::read_header(header1_stream, header, 0));
```
Comment (bob-carpenter) [inline src/test/unit/io/stan_csv_reader_test.cpp]: With 0 passed in, nothing ever gets written, so that branch doesn't get tested.

Add the exact same test with a copy of the header stream and a stringstream ptr for output.

Extra credit for testing the output strings.

File: src/test/unit/io/stan_csv_reader_test.cpp [hunk 5]
```diff
@@ -524,12 +526,14 @@ TEST_F(StanIoStanCsvReader,ParseEpil) {
...
   EXPECT_FLOAT_EQ(6.02445, epil0.timing.sampling);
-  
+
+  EXPECT_EQ("", out.str());
 }
 
 TEST_F(StanIoStanCsvReader,ParseBlockerNondiag) {
   stan::io::stan_csv blocker_nondiag;
-  blocker_nondiag = stan::io::stan_csv_reader::parse(blocker_nondiag0_stream);
+  std::stringstream out;
```
Comment (bob-carpenter) [inline src/test/unit/io/stan_csv_reader_test.cpp]: test 0 input, too

File: src/test/unit/mcmc/chains_test.cpp [hunk 6]
```diff
@@ -31,7 +29,9 @@ class McmcChains : public testing::Test {
 };
 
 TEST_F(McmcChains, constructor) {
-  stan::io::stan_csv blocker1 = stan::io::stan_csv_reader::parse(blocker1_stream);
+  std::stringstream out;
+  stan::io::stan_csv blocker1 = stan::io::stan_csv_reader::parse(blocker1_stream, &out);
```
Comment (bob-carpenter) [inline src/test/unit/mcmc/chains_test.cpp]: test 0, too

**Comment by bob-carpenter** [comment]
See the comments on the code.  None of them is super critical, but the highest priority is getting both `0` and `&stringstream` inputs to make sure neither way segfaults.

File: src/test/unit/io/mcmc_writer_test.cpp (hunk 3 above)
Comment (syclik) [inline src/test/unit/io/mcmc_writer_test.cpp]: seed takes an `Eigen::VectorXd`

**Comment by syclik** [comment]
@bob-carpenter, I went through and made the code changes. I also added a lot of tests checking the behavior with both a stringstream and a null stream.

**Comment by bob-carpenter** [comment]
Great.  Thanks.
